### PR TITLE
APPS-3298 Add Messaging Capabilities

### PR DESCRIPTION
**Pull request by bergieboy**
Screen Shot 2022-01-19 at 2 21 56 PM
Screen Shot 2022-01-19 at 2 22 09 PM
Screen Shot 2022-01-19 at 2 22 42 PM
Screen Shot 2022-01-19 at 2 22 52 PM
Screen Shot 2022-01-19 at 2 23 18 PM

### CODE REVIEW

File: src/containers/message/components/ParticipantSelectionTable.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+} :{
+  handleOnSelect :(participants :Map) => void;
+  participants :Map;
+  targetParticipants :Map;
+}) => {
+
+  const participantOptions = participants
+    .valueSeq().filter((participant) => (participant.getIn([ENROLLMENT_STATUS, 0]) !== EnrollmentStatuses.ENROLLED))
+    .map((participant) => ({ id: participant.getIn([PERSON_ID, 0], '') })).toJS();
+
```
Comment (anzioka) [inline src/containers/message/components/ParticipantSelectionTable.js]: this filters out enrolled participants. I think it makes sense to send links to participants who are marked as active/enrolled instead

File: src/containers/message/components/EnterContactInformationForm.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      mutableMap.set('phone', phoneNumber);
+      mutableMap.set('error', !isValidPhoneNumber(phoneNumber, 'US'));
+    });
+    setTargetParticipants(targetParticipants.set(name, contact));
+  };
+
+  return (
+    <>
+      <Grid container justify="center">
+        <GridItem item xs={6}>
```
Comment (UnsungHero97) [inline src/containers/message/components/EnterContactInformationForm.js]: You can modify `GridItem` to always have `item`:

```
const GridItem = styled(Grid).attrs({ item: true })`
  ...
`;

<GridItem xs={6}>
```

File: src/containers/message/components/EnterContactInformationForm.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+const { NEUTRAL } = Colors;
+
+const List = styled.ul`
+  list-style-type: none;
+  padding: 0;
+`;
+
+const ListItem = styled.li`
+  padding-bottom: 15px;
+`;
```
Comment (UnsungHero97) [inline src/containers/message/components/EnterContactInformationForm.js]: MUI offers list components, please use them.

https://v4.mui.com/components/lists/

File: src/containers/message/components/ParticipantSelectionTable.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  targetParticipants
+} :{
+  handleOnSelect :(participants :Map) => void;
+  participants :Map;
+  targetParticipants :Map;
+}) => {
+
+  const participantOptions = participants
+    .valueSeq().filter((participant) => (participant.getIn([ENROLLMENT_STATUS, 0]) !== EnrollmentStatuses.ENROLLED))
+    .map((participant) => ({ id: participant.getIn([PERSON_ID, 0], '') })).toJS();
```
Comment (UnsungHero97) [inline src/containers/message/components/ParticipantSelectionTable.js]:
```suggestion
  const participantOptions = participants
    .valueSeq()
    .filter((participant) => (participant.getIn([ENROLLMENT_STATUS, 0]) !== EnrollmentStatuses.ENROLLED))
    .map((participant) => ({ id: participant.getIn([PERSON_ID, 0], '') }))
    .toJS();
```

File: src/containers/message/components/ParticipantSelectionTable.js (hunk 1 above)
Comment (UnsungHero97) [reply src/containers/message/components/ParticipantSelectionTable.js]: Is the filter even necessary? I think we should just let the user pick which participants to send the text to, no filtering. We can add this if users request it.

File: src/containers/message/sagas/sendMessage.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    yield put(sendMessage.finally(action.id));
+  }
+}
+
+function* sendMessageWatcher() :Saga<*> {
+  yield takeEvery(SEND_MESSAGE, sendMessageWorker);
+}
+
+export {
+  sendMessageWatcher
```
Comment (UnsungHero97) [inline src/containers/message/sagas/sendMessage.js]: Missing worker export, though it's not necessary here.

File: src/core/sagas/Sagas.js [hunk 6]
```diff
@@ -68,6 +69,9 @@ export default function* sagas() :Generator<*, *, *> {
     fork(QuestionnaireSagas.getStudyQuestionnairesWatcher),
     fork(QuestionnaireSagas.submitQuestionnaireWatcher),
 
+    // permissions
```
Comment (UnsungHero97) [inline src/core/sagas/Sagas.js]:
```suggestion
    // messages
```

File: src/containers/message/sagas/sendMessage.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+const LOG = new Logger('MessageSagas');
+
+function* sendMessageWorker(action :SequenceAction) :Saga<*> {
+  try {
+    yield put(sendMessage.request(action.id));
+    const { messageDetailsList, organizationId } = action.value;
+
+    const response = yield call(ChronicleApi.sendMessage, organizationId, messageDetailsList);
+    if (response.error) throw response.error;
```
Comment (UnsungHero97) [inline src/containers/message/sagas/sendMessage.js]: This project is using axios directly which means `response.error` doesn't exist. This is what a response object looks like:

https://github.com/axios/axios/#response-schema

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

File: src/containers/message/components/ParticipantSelectionTable.js (hunk 1 above)
Comment (bergieboy) [reply src/containers/message/components/ParticipantSelectionTable.js]: you want to send enrollment links to people that are already enrolled?

Comment (anzioka) [reply src/containers/message/components/ParticipantSelectionTable.js]: 'Enrolled' status doesn't mean that the participant has enrolled their device. In the current usage it implies that the participant is active (they are able to participate in the data collection process)

Comment (UnsungHero97) [reply src/containers/message/components/ParticipantSelectionTable.js]: No. I want to let users do what they want and not stand in the way. Filtering will prevent them from choosing certain participants and I guarantee you there will be a support ticket that says "I cannot text participant x".

**Review by solomontang — APPROVED** [review]

**Review by anzioka — APPROVED** [review]

**Review by UnsungHero97 — APPROVED** [review]

**Pull request closed without merging**